Framework/Parallel/inc/MantidParallel/Communicator.h

```diff
@@ -0,0 +1,95 @@
+#ifndef MANTID_PARALLEL_COMMUNICATOR_H_
+#define MANTID_PARALLEL_COMMUNICATOR_H_
+
+#include "MantidParallel/DllConfig.h"
+#include "MantidParallel/Request.h"
+#include "MantidParallel/ThreadingBackend.h"
+
+#include <boost/make_shared.hpp>
+#include <boost/mpi/communicator.hpp>
+
+namespace Mantid {
+namespace Parallel {
+class ParallelRunner;
+
+/** Wrapper for boost::mpi::communicator. For non-MPI builds an equivalent
+  implementation is provided.
+
+  @author Simon Heybrock
+  @date 2017
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_PARALLEL_DLL Communicator {
+public:
+  Communicator() = default;
+  explicit Communicator(const boost::mpi::communicator &comm);
+
+  int rank() const;
+  int size() const;
+  template <typename... T> void send(T &&... args) const;
+  template <typename... T> void recv(T &&... args) const;
+  template <typename... T> Request isend(T &&... args) const;
+  template <typename... T> Request irecv(T &&... args) const;
+
+private:
+  Communicator(boost::shared_ptr<detail::ThreadingBackend> backend,
+               const int rank);
+
+  boost::mpi::communicator m_communicator;
+  boost::shared_ptr<detail::ThreadingBackend> m_backend;
```

Comment:
It can only be initialized by a private constructor, accessible by `ParallelRunner`. This is not intented for generic use but only unit tests, so I do not provide another way for access.